nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -8,21 +8,183 @@ use super::DataStore;
+        &self,
+        log: &Logger,
+        desired_version: SemverVersion,
+        config: Option<&SchemaConfig>,
+    ) -> Result<(), String> {
+        let mut current_version = match self.database_schema_version().await {
+            Ok(current_version) => {
+                // NOTE: We could run with a less tight restriction.
+                //
+                // If we respect the meaning of the semver version, it should be possible
+                // to use subsequent versions, as long as they do not introduce breaking changes.
+                //
+                // However, at the moment, we opt for conservatism: if the database does not
+                // exactly match the schema version, we refuse to continue without modification.
+                if current_version == desired_version {
+                    info!(log, "Compatible database schema: {current_version}");
+                    return Ok(());
+                }
+                let observed = &current_version.0;
+                warn!(log, "Incompatible database schema: Saw {observed}, expected {desired_version}");
+                current_version
+            }
+            Err(e) => {
+                return Err(format!("Cannot read schema version: {e}"));
+            }
+        };
+
+        let Some(config) = config else {
+            return Err("Not configured to automatically update schema".to_string());
+        };
+
+        if current_version > desired_version {
+            return Err("Nexus older than DB version: automatic downgrades are unsupported".to_string());
+        }
+
+        // If we're here, we know the following:
+        //
+        // - The schema does not match our expected version (or at least, it
+        // didn't when we read it moments ago).
+        // - We should attempt to automatically upgrade the schema.
+        //
+        // We do the following:
+        // - Look in the schema directory for all the changes, in-order, to
+        // migrate from our current version to the desired version.
+
+        info!(log, "Reading schemas from {}", config.schema_dir.display());
+        let mut dir = tokio::fs::read_dir(&config.schema_dir)
+            .await
+            .map_err(|e| format!("Failed to read schema config dir: {e}"))?;
+        let mut all_versions = BTreeSet::new();
+        while let Some(entry) = dir
+            .next_entry()
+            .await
+            .map_err(|e| format!("Failed to read schema dir: {e}"))?
+        {
+            if entry.file_type().await.map_err(|e| e.to_string())?.is_dir() {
+                let name = entry
+                    .file_name()
+                    .into_string()
+                    .map_err(|_| "Non-unicode schema dir".to_string())?;
+                if let Ok(observed_version) = name.parse::<SemverVersion>() {
+                    all_versions.insert(observed_version);
+                } else {
+                    warn!(log, "Failed to parse {name} as a semver version");
```

Comment:
do we want to be more conservative here? I get that we may not want to detonate if someone left behind a temp file or something... but is this surprising enough that we might want to refuse to proceed?

Reply:
Sure, I'll throw an error here.